lib/puttyhosts/puttyhosts.go

```diff
@@ -238,3 +238,42 @@ func FormatHostCAPublicKeysForRegistry(hostCAPublicKeys map[string][]string, hos
 	}
 	return registryOutput
 }
+
+// SplitValidityKey processes PuTTY's "Validity" string key into individual list elements.
+// PuTTY uses a custom string format to represent what hostnames a given key should be trusted for.
+// See https://the.earth.li/~sgtatham/putty/0.79/htmldoc/Chapter4.html#config-ssh-cert-valid-expr for details.
+func SplitValidityKey(input string) []string {
+	var output []string
+	var stack []rune
+	var currentToken string
+
+	// Iterate over the input string and split out the hostnames inside it on the logical OR token "||"
+	// If any values are encountered inside parentheses (), ignore them and pass them through as-is.
+	// This is done to avoid implementing a full lexer as in https://git.tartarus.org/?p=simon/putty.git;a=blob;f=utils/cert-expr.c
+	// because we only care about the individual host elements inside.
+	for i := 0; i < len(input); i++ {
+		char := rune(input[i])
+
+		if char == '(' {
+			stack = append(stack, char)
+			currentToken += string(char)
+		} else if char == ')' {
+			if len(stack) > 0 {
+				stack = stack[:len(stack)-1]
+				currentToken += string(char)
+			}
+		} else if i <= len(input)-2 && input[i:i+2] == "||" && len(stack) == 0 {
+			output = append(output, strings.TrimSpace(currentToken))
+			currentToken = ""
+			i += 1 // skip past the other character in the token
+		} else {
+			currentToken += string(char)
+		}
+	}
+
+	if len(currentToken) > 0 {
+		output = append(output, strings.TrimSpace(currentToken))
+	}
```

Comment:
This works well but seems a little gross, so any feedback on a tidier/more efficient way to implement it is gratefully received.

Reply:
@nklaassen is very good at parsing/validating this type of expression language. Worth checking with him if there's anything we might be able to reuse or at least follow as an example.

Reply:
Lexer code makes me squirm, it's so hard to get right. As is, this seems like it could behave badly (generate invalid expressions) when there are other operators (`&&` or `!`) or unmatched parens present.

I have really only been building stuff with gravitational/predicate which uses go/parser under the hood, which I don't think would handle this syntax. Is there any way it's possible to avoid parsing this at all? Could we make it write-only, completely regenerated each time we need to?

If we can't avoid reading it completely, could we only touch it if we can make sure it strictly follows the pattern `a || b || c` by:
  * first splitting on whitespace
  * ensure there's an odd number of tokens
  * ensure every odd indexed token is exactly `||`
  * ensure there is no `&&`, `!` or parens anywhere

If this all passes I think we could safely add an `|| hostname` at the end. I'm thinking/hoping it's not necessary to support parsing expressions after users have been messing with them too much.

Reply:
The test cases work OK with `&&` and matched parens, but I haven't tested `!`. I did also test this earlier with mismatched parens, as well as using `|` instead of `||` and you're right, it does break in those cases.

Unfortunately there's no way to make this both functional _and_ write-only - each time a new host is added, it's essential that we preserve all the hosts which were previously added (hopefully always by `tsh puttyconfig`) and then append the new host to the end. The only alternative would be to try to re-read and parse every saved Teleport session from PuTTY's saved session list, which would itself introduce considerable instability.

Your feedback has given me some good ideas for how to make this more resilient, though. I think there are two good options:

1) Add additional validation so that we don't operate on `Validity` strings which are obviously broken. PuTTY is going to reject a non-parsing string when attempting to connect anyway, so there's no real change in functionality. Some additional checks we could do:
- check that the count of `(` is equal to the count of `)`, and error out if not
- check that every instance of `|` is immediately followed by `|`, and error out if not
- check that every instance of `&` is immediately followed by `&`, and error out if not
- error out if the string contains `!` or `:`, as Teleport has no need to make use of this functionality and we don't need to support others introducing it.

2) Reimplement this exactly as you described above and just error out if the string doesn't match that `<value> || <value> || ...` pattern. I don't see any need for us to actually support people modifying the validity outside of Teleport, and if they do it's fair to say they should expect that things will break.

The only real downside with both of these approaches is that it'll make us more likely to have to show errors to the end user, but I can add an additional steps to the troubleshooting section of the documentation detailing how to manually clear a broken Validity list. I could also potentially implement a `tsh puttyconfig --clear-host-cas` command for less technical users which purges all Teleport-managed Host CAs from the registry and starts afresh, although I'd prefer not to creep the scope of this PR in the hope that I can merge this for 14.0.1, or 14.0.2 at the latest.

Reply:
This has been reimplemented using option 2, and I have added a troubleshooting step to the docs with instructions on how to fix any resulting validity errors.